src/Helper/Quote.php

```diff
@@ -109,16 +109,42 @@ public function __construct(
     public function getValidatedCoupons()
     {
         $coupons = [];
+        $returnInvalidCartRuleCoupon = false;
+        $amount = $this->quote->getSubtotal() - $this->quote->getSubtotalWithDiscount();
+        $couponsIncludeTax = $this->taxHelper->couponInclTax();
         foreach ($this->sgBase->getExternalCoupons() as $coupon) {
+            if ($coupon->getCode() == '1'){
+                $returnInvalidCartRuleCoupon = true;
+                continue;
+            }
             if (!$coupon->getNotValidMessage()) {
-                $amount = $this->quote->getSubtotal() - $this->quote->getSubtotalWithDiscount();
                 $coupon->setIsValid(true);
                 $coupon->setCurrency($this->quote->getStoreCurrencyCode());
                 $coupon->setIsFreeShipping((bool) $this->quote->getShippingAddress()->getFreeShipping());
-                $this->taxHelper->couponInclTax() ? $coupon->setAmountGross($amount) : $coupon->setAmountNet($amount);
+                $couponsIncludeTax ? $coupon->setAmountGross($amount) : $coupon->setAmountNet($amount);
             }
             $coupons[] = $this->couponHelper->dataToEntity($coupon->toArray());
         }
+        if (empty($coupons) && !empty($amount)) {
+            $coupon = new \ShopgateExternalCoupon();
+            $coupon->setCode('1');
+            $coupon->setIsValid(true);
+            $coupon->setCurrency($this->quote->getStoreCurrencyCode());
+            $coupon->setIsFreeShipping((bool) $this->quote->getShippingAddress()->getFreeShipping());
+            $couponsIncludeTax ? $coupon->setAmountGross($amount) : $coupon->setAmountNet($amount);
```

Comment:
there seems to be a lot of copy pasting here, can you refactor this logic to a function by chance?

Reply:
how is it now?

Reply:
are you saying you are done and I can review?

Reply:
Sorry for the confusion. No.
I will leave a comment that I am done with the feedback when I am ready.